third_party/move/move-compiler-v2/src/pipeline/control_flow_graph_simplifier.rs

```diff
@@ -0,0 +1,599 @@
+            let succ_block = self.get_the_non_trivial_successor(block);
+            if let Some(next_to_vist) = next_block_to_visit {
+                if *next_to_vist == succ_block {
+                    remove_tail_jump(&mut code_block);
+                }
+            }
+        }
+        code_block
+    }
+
+    /// Checks whether a block falls to the next block without jump, branch, abort, or return;
+    /// i.e., the block is followed by the next in the original code
+    fn falls_to_next_block(code: &[Bytecode]) -> bool {
+        let last_instr = code.last().expect("last instr");
+        !last_instr.is_always_branching()
+    }
+
+    /// Gets the only successor of `block`; panics if there is no successors.
+    /// (May not panic in the release version due to the use of debug_assert)
+    fn get_the_successor(&self, block: BlockId) -> BlockId {
+        let successors = self.successors.get(&block).expect("successors");
+        debug_assert!(successors.len() == 1);
+        *successors.first().expect("successor block")
+    }
+
+    /// Gets the only successor of `block` which is not entry/exit block; panics if this is not the case.
+    /// (May not panic in the release version due to the use of debug_assert)
+    fn get_the_non_trivial_successor(&self, block: BlockId) -> BlockId {
+        let the_suc = self.get_the_successor(block);
+        debug_assert!(!self.is_trivial_block(the_suc));
+        the_suc
+    }
+
+    /// Adds an explicit jump to `to_block` to the end of `code`
+    fn add_explicit_jump(&self, code: &mut Vec<Bytecode>, to_block: BlockId) {
+        debug_assert!(!self.is_trivial_block(to_block));
+        let to_label = self.get_block_label(to_block).expect("label");
+        let attr_id = code.last().expect("instruction").get_attr_id();
+        code.push(Bytecode::Jump(attr_id, to_label));
+    }
+
+    /// Returns the instructions of the block
+    fn block_instrs(&self, block_id: BlockId) -> &[Bytecode] {
+        self.code_blocks.get(&block_id).expect("block instructions")
+    }
+
+    /// Returns the label of the block or `None` if it doesn't start with a label
+    fn get_block_label(&self, block_id: BlockId) -> Option<Label> {
+        if let Bytecode::Label(_, label) = self
+            .block_instrs(block_id)
+            .first()
+            .expect("first instruction")
+        {
+            Some(*label)
+        } else {
+            None
+        }
+    }
+
+    /// Checks if the block is entry or exit block
+    fn is_trivial_block(&self, block: BlockId) -> bool {
+        block == self.entry_block || block == self.exit_block
+    }
+
+    /// Removes a block from the control flow graph
+    /// `pred_action`: action to take for each predecessor of the block
+    /// `succ_action`: action to take for each successor of the block
+    fn remove_block<F, G>(&mut self, block_to_remove: BlockId, pred_action: F, succ_action: G)
+    where
+        F: FnOnce(&mut Self, BlockId) + Copy,
```

Comment:
Doesn't `FnOnce` mean the function can only be called once?  What happens if there are multiple predecessors?

Reply:
It will be copied. But I'll rethink if this should be a `FnOnce`.